src/saml2/assertion.py

```diff
@@ -276,162 +277,122 @@ def restriction_from_attribute_spec(attributes):
-        if "sp_entity_id" in kwargs:
-            ecs = kwargs["mds"].entity_categories(kwargs["sp_entity_id"])
-            for ec_map in maps:
-                for key, (atlist, only_required) in ec_map.items():
-                    if key == "":  # always released
-                        attrs = atlist
-                    elif isinstance(key, tuple):
-                        if only_required:
-                            attrs = [a for a in atlist if a in required]
-                        else:
-                            attrs = atlist
-                        for _key in key:
-                            if _key not in ecs:
-                                attrs = []
-                                break
-                    elif key in ecs:
-                        if only_required:
-                            attrs = [a for a in atlist if a in required]
-                        else:
-                            attrs = atlist
-                    else:
-                        attrs = []
+    :param restrictions: policy configuration
+    :return: The assertion with the string specification replaced with
+        a compiled regular expression.
+    """
+    for who, spec in restrictions.items():
+        spec = spec or {}
 
-                    for attr in attrs:
-                        restrictions[attr] = None
-        else:
-            for ec_map in maps:
-                for attr in ec_map[""]:
-                    restrictions[attr] = None
+        entity_categories = spec.get("entity_categories", [])
+        ecs = []
+        for cat in entity_categories:
+            try:
+                _mod = importlib.import_module(cat)
+            except ImportError:
+                _mod = importlib.import_module("saml2.entity_category.%s" % cat)
+
+            _ec = {}
+            for key, items in _mod.RELEASE.items():
+                alist = [k.lower() for k in items]
+                _only_required = getattr(_mod, "ONLY_REQUIRED", {}).get(key, False)
+                _ec[key] = (alist, _only_required)
+            ecs.append(_ec)
+        spec["entity_categories"] = ecs or None
+
+        attribute_restrictions = spec.get("attribute_restrictions") or {}
+        _attribute_restrictions = {}
+        for key, values in attribute_restrictions.items():
+            lkey = key.lower()
+            values = [] if not values else values
+            _attribute_restrictions[lkey] = (
+                [re.compile(value) for value in values] or None
+            )
+        spec["attribute_restrictions"] = _attribute_restrictions or None
 
     return restrictions
 
 
 class Policy(object):
-    """ handles restrictions on assertions """
+    """Handles restrictions on assertions."""
 
-    def __init__(self, restrictions=None):
-        if restrictions:
-            self.compile(restrictions)
-        else:
-            self._restrictions = None
+    def __init__(self, restrictions=None, mds=None):
+        self.metadata_store = mds
```

Comment:
Policy needs a metadata store to look up attributes (entity-categories, registration-authority, etc).

Previously `filter`, `restrict`, and `get_entity_categories` requested the metadata store (aka `mds`, `mdstore`) as a param. Now, the one that has been supplied during initialization is used.

Reply:
There are many changes, but this is the core of it.

To make this possible, the config module changed to build a Policy object with the metadata store. The metadata store must have been loaded beforehand.

Once the Policy object got a metadata-store object, the interfaces were refactored to remove the explicit param. The tests where then adjusted.